File: Week_02/java/src/graph/_733_FloodFill.java
package graph;

import java.util.LinkedList;

class _733_FloodFill {
    int[][] dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
    //时间O(m * n) 空间O(m * n)
    public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
        int oldColor = image[sr][sc];
        if (oldColor == newColor) return image;
        int m = image.length;
        int n = image[0].length;
        dfs(image, sr, sc, newColor, oldColor);
        return image;
    }

    void dfs(int[][] image, int i, int j, int newColor, int oldColor) {
        if (i < 0 || i >= image.length || j < 0 || j >= image[0].length || image[i][j] != oldColor) {
            return;
        }
        image[i][j] = newColor;
        for (int[] dir : dirs) {
            dfs(image, i + dir[0], j + dir[1], newColor, oldColor);
        }
    }

    class SolutionBfs {
        //时间O(m * n) 空间O(m * n)
        public int[][] floodFill(int[][] image, int sr, int sc, int newColor) {
            int oldColor = image[sr][sc];
            if (oldColor == newColor) {
                return image;
            }
            int m = image.length;
            int n = image[0].length;
            LinkedList<int[]> queue = new LinkedList<>();
            queue.offer(new int[]{sr, sc});
            int[][] dirs = {{1, 0}, {-1, 0}, {0, 1}, {0, -1}};
            while (!queue.isEmpty()) {
                int[] p = queue.poll();
                int i = p[0], j = p[1];
                image[i][j] = newColor;
                for (int[] dir : dirs) {
                    int in = i + dir[0];
                    int jn = j + dir[1];
                    if (in >= 0 && in < m && jn >= 0 && jn < n && image[in][jn] == oldColor) {
                        queue.offer(new int[]{in, jn});
                    }
                }
            }
            return image;
        }
    }
}

File: Week_02/kotlin/src/graph/_733_FloodFill.kt
package graph

import java.util.*

fun floodFill(image: Array<IntArray>, sr: Int, sc: Int, newColor: Int): Array<IntArray> {
    val oldColor = image[sr][sc]
    if (oldColor == newColor) return image
    val m = image.size
    val n = image[0].size
    val queue = LinkedList<IntArray>()
    val dirs = arrayOf(intArrayOf(1, 0), intArrayOf(-1, 0), intArrayOf(0, 1), intArrayOf(0, -1))
    queue.offer(intArrayOf(sr, sc))
    while (queue.isNotEmpty()) {
        val (i, j) = queue.poll()
        image[i][j] = newColor
        for (dir in dirs) {
            val x = i + dir[0]
            val y = j + dir[1]
            if (x in 0 until m && y in 0 until n && image[x][y] == oldColor) {
                queue.offer(intArrayOf(x, y))
            }
        }
    }
    return image
}

class _733_Dfs {
    fun floodFill(image: Array<IntArray>, sr: Int, sc: Int, newColor: Int): Array<IntArray> {
        val oldColor = image[sr][sc]
        if (oldColor == newColor) return image
        val m = image.size
        val n = image[0].size
        val dirs = arrayOf(intArrayOf(1, 0), intArrayOf(-1, 0), intArrayOf(0, 1), intArrayOf(0, -1))
        fun dfs(i: Int, j: Int) {
            if (i !in 0 until m || j !in 0 until n || image[i][j] != oldColor) {
                return
            }
            image[i][j] = newColor
            for (dir in dirs) {
                dfs(i + dir[0], j + dir[1])
            }
        }
        dfs(sr, sc)
        return image
    }
}